locale/en/knowledge/javascript-conventions/how-to-create-default-parameters-for-functions.md

```diff
@@ -8,61 +8,74 @@ difficulty: 1
-
-    var Args = require("vargs").Constructor;
-
-    example = function () {
-      var args = new Args(arguments);
-      args.callback.apply({},args.all);
-    }
-
-    example("The first parameter", console.log);
-    example("The first parameter", "and second parameter", console.log);
-    example("The first parameter", "and second parameter", "and third parameter", console.log);
-    example("The first parameter", "and second parameter", "and third parameter", "etc", console.log);
+Usually a function will take a set number of parameters, and require that all of them be present before it can be executed successfully. However, you will sometimes run into situations where you want to provide a default value for a parameter or take a variable number of parameters. Fortunately, from ES6/ES2015 **default parameters** is in the language specification:
+
+```js
+const pow = (base, power = 2) => {
+  return Math.pow(base, power);
+}
+
+console.log(pow(2)); // 4
+console.log(pow(2,10)); // 1024
+```
+In the above code The function `pow` return square of a number or any other power specified in the function call because the argument `power` is given a default value of 2 so whenever no second argument is provided or the provided value is `undefined` the function `pow` will use 2 as the value of argument `power`. But there is a small gotcha in it:
+
+```js
+const pow = (base, power = 2) => {
+  return Math.pow(base, power);
+}
+
+console.log(pow(2, undefined)); // 4
+console.log(pow(2, null)); // 1
+```
+In JavaScript there are many ways of indicating ["falsy"(false values)](/en/knowledge/javascript-conventions/what-are-truthy-and-falsy-values/) but out of them, only `undefined` will trigger the default parameter. This is the reason why `pow(2, undefined)` returns 4 and `pow(2, null)` returns 1.
+
+JS PRO TIP: In `Math.pow(base, power)` if `power` is 0 or any other "falsy" value (except NaN) the result will always be 1.
+
+Unfortunately, previous versions of javascript (ES5 and below) does not support **default parameters** out of the box but over time, however, people have developed idioms to compensate.
+
+The first idiom is giving a default value for the last parameter. This is done by checking if the last parameter is `undefined` and setting it to a default value if it is. Sometimes people use the idiom: `optionalParameter = optionalParameter || defaultValue`. This can have some undesirable behaviour when they pass values that are equal to false such as `false`, `0`, and `""`. So a better way to do this is by explicitly checking that the optional parameter is `undefined`. Here is some code showing the two styles and the differing behaviour:
+
+```js
+const example = function (optionalArg) {
+  optionalArg = optionalArg || "No parameter was passed";
+  console.log(optionalArg);
+}
+
+const betterExample = function (optionalArg) {
+  if (typeof optionalArg === 'undefined') {
```

Comment:
What about using `optionalArg === undefined` directly instead of using `typeof` to check the type?

Reply:
Thanks, @Maledong for that great input 🙂
Though in these examples particularly, it won't make any difference using `optionalArg === undefined` or `typeof optionalArg === 'undefined'` but in general `typeof` is slightly more reliable/predictable.
This answer [here](https://stackoverflow.com/a/4726198/9991821) on StackOverflow provides a great explanation on this.

I would love to have your suggestions/further discussion on this

Reply:
@Yash-Handa：I see, but the answer means if you cannot decide whether your variable is defined or not, you can safely use `typeof somevariable === 'undefined'`, since in our senario, we already have the variable defined `optionalArg` as the input parameter, and it isn't a dynamic one. So I tend to use `optionalArg === undefined`. :)

Reply:
Yeah @Maledong, In these examples it won't matter if we use `optionalArg === undefined` or `typeof optionalArg === 'undefined'`.
As you suggest, so let us go with `optionalArg === undefined` approach. It would simplify (less verbose) the code. 😎